File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        // TC: O(n^3)
        // SC: O(n) n is the length of String s
        
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) { // O(n)
            for (int j = 0; j < i; j++) { // O(n)
                if (dp[j] && wordDict.contains(s.substring(j, i))) { // O(n)
                    dp[i] = true;
                    break;
                }
            }
        }

        // <= 10^6
        return dp[s.length()];
    }
}

File: 0139-word-break/0139-word-break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    const n = s.length;
    const dp = new Array(n+1).fill(false);
    dp[0] = true;
    wordDict = new Set(wordDict);

    for (let i = 1; i <= n; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && wordDict.has(s.slice(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[n];
};